credentials/credentials_test.go

```diff
@@ -59,3 +63,104 @@ func TestTLSClone(t *testing.T) {
 		t.Fatalf("Change in clone should not affect the original, c.Info().ServerName = %v, want %v", c.Info().ServerName, expectedServerName)
 	}
 }
+
+const tlsDir = "../test/testdata/"
+
+func TestTLSClientHandshakeReturnsAuthInfo(t *testing.T) {
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Failed to listen: %v", err)
+	}
+	defer lis.Close()
+	serverTLS, err := NewServerTLSFromFile(tlsDir+"server1.pem", tlsDir+"server1.key")
+	if err != nil {
+		t.Fatalf("Failed to create server TLS. Error: %v", err)
+	}
+	var serverAuthInfo TLSInfo
+	errChan := make(chan error)
+	go func() {
+		var sErr error
+		defer func() {
+			errChan <- sErr
+		}()
+		serverRawConn, sErr := lis.Accept()
+		if sErr != nil {
+			t.Errorf("Server failed to accept connection: %v", sErr)
+			return
+		}
+		serverConn := tls.Server(serverRawConn, serverTLS.(*tlsCreds).config)
+		sErr = serverConn.Handshake()
+		if sErr != nil {
+			t.Errorf("Error on server while handshake. Error: %v", sErr)
+			return
+		}
+		serverAuthInfo = TLSInfo{serverConn.ConnectionState()}
+	}()
+	conn, err := net.Dial("tcp", lis.Addr().String())
+	if err != nil {
+		t.Fatalf("Client failed to connect to %v. Error: %v", lis.Addr().String(), err)
+	}
+	defer conn.Close()
+	c := NewTLS(&tls.Config{InsecureSkipVerify: true})
+	_, authInfo, err := c.ClientHandshake(context.Background(), lis.Addr().String(), conn)
+	if err != nil {
+		t.Fatalf("Error on client while handshake. Error: %v", err)
+	}
+	// wait until server has populated the serverAuthInfo struct.
+	if err = <-errChan; err != nil {
+		return
+	}
```

Comment:
and the above client side logic can be wrapped into a function too:

`func clientHandle(hs func(context.Context, string, net.Conn) (net.Conn, AuthInfo, error))`